gcputil/gcputil.go

```diff
@@ -0,0 +1,120 @@
+// the stack driver exporter, if exists in the underlying platform.
+// If exporter is registered, it returns the exporter's flush function so you can
+// ensure all metrics are uploaded. The flush function is safe to call.
+func RegisterOpenCensus(projectID string, onErr func(error)) (flush func(), err error) {
+	svcName, svcVersion := "", ""
+	if IsGAE() {
+		_, svcName, svcVersion = GetGAEInfo()
+	} else if n, v := os.Getenv("SERVICE_NAME"), os.Getenv("SERVICE_VERSION"); n != "" {
+		svcName, svcVersion = n, v
+	}
+	opts := SDExporterOptions(projectID, svcName, svcVersion, onErr)
+	if opts == nil {
+		return func() {}, nil
+	}
+	exp, err := initSDExporterV2(*opts)
+	if err != nil {
+		return func() {}, nil
+	}
+	return exp.Flush, nil
+}
+
+// GoogleProjectID returns the GCP Project ID
+// that can be used to instantiate various
+// GCP clients such as Stack Driver.
+func GoogleProjectID() string {
+	return os.Getenv("GOOGLE_CLOUD_PROJECT")
+}
+
+// IsGAE tells you whether your program is running
+// within the App Engine platform.
+func IsGAE() bool {
+	return os.Getenv("GAE_DEPLOYMENT_ID") != ""
+}
+
+// GetGAEInfo returns the GCP Project ID,
+// the service, and the version of the application.
+func GetGAEInfo() (projectID, service, version string) {
+	return GoogleProjectID(),
+		os.Getenv("GAE_SERVICE"),
+		os.Getenv("GAE_VERSION")
+}
+
+// SDExporterOptions returns Stack Driver Options that you can pass directly
+// to the OpenCensus exporter or other libraries.
+func SDExporterOptions(projectID, service, version string, onErr func(err error)) *stackdriver.Options {
+	var mr monitoredresource.Interface
+	if m := monitoredresource.Autodetect(); m != nil {
+		mr = m
+	} else if IsGAE() {
+		mr = gaeInterface{
+			typ: "gae_app",
+			labels: map[string]string{
+				"project_id": projectID,
+			},
+		}
+	}
+	if mr == nil {
+		return nil
+	}
+
+	return &stackdriver.Options{
+		ProjectID:               projectID,
+		MonitoredResource:       mr,
+		OnError:                 onErr,
+		DefaultMonitoringLabels: &stackdriver.Labels{},
+		DefaultTraceAttributes: map[string]interface{}{
+			"service": service,
+			"version": version,
+		},
+	}
+}
+
+// InitSDExporter will initialize the OpenCensus tracing/metrics exporter
+func InitSDExporter(opts stackdriver.Options) error {
+	exporter, err := stackdriver.NewExporter(opts)
+	if err != nil {
+		return err
+	}
+	trace.RegisterExporter(exporter)
+	view.RegisterExporter(exporter)
+	return nil
+}
+
+// TODO(marwan): this is so that I don't break server/kit
+// see https://github.com/NYTimes/gizmo/issues/193
+func initSDExporterV2(opts stackdriver.Options) (*stackdriver.Exporter, error) {
+	exporter, err := stackdriver.NewExporter(opts)
+	if err != nil {
+		return nil, err
+	}
+	trace.RegisterExporter(exporter)
+	view.RegisterExporter(exporter)
+	return exporter, nil
+}
+
+// implements contrib.go.opencensus.io/exporter/stackdriver/monitoredresource.Interface
+type gaeInterface struct {
```

Comment:
I've been meaning to remove this one and replace it with the same impl we use for the sd logger. Mind removing this and replacing it with the same "google.golang.org/genproto/googleapis/api/monitoredres" implementation?

Reply:
@jprobinson I don't see the `MonitoringResource()` method here https://godoc.org/google.golang.org/genproto/googleapis/api/monitoredres#MonitoredResource

Do you mind pointing me to the exact code where the interface is implemented? I'll switch it up once I see it, thanks

Reply:
It's a bit chunked up but it's initiated here:  https://github.com/NYTimes/gizmo/blob/master/server/kit/sd_log.go#L39-L45

and the `gae_app` type is set after the `isGAE` check: https://github.com/NYTimes/gizmo/blob/master/server/kit/sd_log.go#L47

Reply:
@jprobinson I don't think this works for StackDriver Exporter because the Resource field is deprecated https://github.com/census-ecosystem/opencensus-go-exporter-stackdriver/blob/master/stackdriver.go#L130

And the monitoredresource does not implement Interface

Reply:
Also, the AutoDetect does not have GAE, only GKE, GCE and AWS...maybe GAE does not need one at all?

Reply:
It does. I imagine since the GAE 2nd gen runtime is in beta, maybe they haven't considered adding it to OC yet, but not sure.

I do know that the OC tracing exporters certainly did not wire themselves up properly until I added some information, at least. The project/service/version labels allow this to match the rest of the GAE paradigms so I figured it was worth adding.